mvebu: add support for RB5009UG+S+IN

robimarko opened the pull request:
This patch adds support for Mikrotik RB5009UG+S+IN.

Specifications:
  - SoC: Marvell Armada 7040 (88F7040) - 4 cores, ARMv8 Cortex-A72, 1.4GHz, 64bit
  - RAM: 1024MB DDR4
  - Flash: 16MB SPI NOR flash, 1024MB NAND
  - Ethernet:
  	* Marvell 88E6393X - Amethyst:
  	* one 2.5G RJ45 port via Qualcomm QCA8081 PHY
  	* seven 1G RJ45 ports via built-in PHY-s
  	* one 10G SFP+ cage
  	* All ports share the same 10G switch uplink to the CPU
  - LED: User, SFP, Hdr1, Hdr2
  - Buttons: Reset
  - UART: 115200 8n1 on the MikroTik 16 pin header
  - USB: One USB3 port
  - Power: 24-57 V via
  	* DC jack
  	* 802.3af/at PoE on Ethernet 1
  	* 2-pin terminal on the side

16 Pin header pinout:
```
1   GND Vcc  RX  ?  GND
   #--------------------#
   |.-. .-. .-. .-. .-. |
   |'-' '-' '-' '-' '-' |
   |.-. .-. .-. .-. .-. |
   |'-' '-' '-' '-' '-' |
   #--------------------#
2   CLK  DO /CS  TX  DI
```

Do note that the default RouterBoot has disabled UART even when the
required hard-config bit is set to indicate UART support.
Patched RouterBoot must be used if UART is desired.

Also, since ARM64 Linux support does not support in any way appending the
DTB to the kernel image we use mainline U-Boot with added RB5009 support
in order to boot OpenWrt.
MikroTik uses YAFFS to store the boot kernel and we use YAFUT to put U-Boot
as the kernel which RouterBoot then simply boots as an ELF.

Install instructions:

NOTE: In case you are using an existing out of tree version of OpenWrt make
sure to reinstall RouterOS via Netinstall to return the expected partition
layout.

1. Prepare FAT or EXT4 formatted USB drive with OpenWrt initramfs:
* Copy bin/targets/mvebu/cortexa72/openwrt-mvebu-cortexa72-mikrotik_rb5009-initramfs-uImage.itb
to the root of FAT or EXT4 formatted USB drive.
* Plug in the drive to the RB5009 USB port

2. Boot the modified OpenWrt built U-Boot ELF:
u-boot.elf from bin/targets/mvebu/cortexa72/u-boot-rb5009/u-boot.elf

Consult OpenWrt wiki for common instructions on switching to boot from
Ethernet once as well as serving the file:
https://openwrt.org/toh/mikrotik/common

Once U-Boot is booted it will attempt to boot in the following order:
1. NAND
2. USB
3. Network

NAND is expected to fail but USB or Networking need to serve the OpenWrt
initramfs image and after booting it will be accessible from LAN ports
on the default 192.168.1.1 IP with default credentials.

3. Flash modified RouterBoot that enables UART (Optional but recommended):
https://public.robimarko.eu/RB5009/70x0-7.15-uart.fwf

* Copy the file over to the booted OpenWrt initramfs to /tmp
* Run: mtd erase RouterBOOT-primary
* Run: mtd write /tmp/70x0-7.15-uart.fwf RouterBOOT-primary

4. Install U-Boot to boot OpenWrt:
* Copy the u-boot.elf from bin/targets/mvebu/cortexa72/u-boot-rb5009/u-boot.elf
to OpenWrt initramfs to /tmp.
* Run: . /lib/functions.sh
* Run: yafut -d /dev/mtd$(find_mtd_index "YAFFS") -w -i /tmp/u-boot.elf -o kernel -T
This will use yafut to copy the U-Boot as kernel in YAFFS so that RouterBoot boots it.

5. Wipe the NAND UBI partition:
* Run: ubiformat /dev/mtd$(find_mtd_index "ubi") -y
This will prepare the existing RouterOS rootfs partition for OpenWrt.

6. Flash OpenWrt:
* Copy the bin/targets/mvebu/cortexa72/openwrt-mvebu-cortexa72-mikrotik_rb5009-squashfs-sysupgrade.bin
to OpenWrt initramfs to /tmp.
* Run: sysupgrade /tmp/openwrt-mvebu-cortexa72-mikrotik_rb5009-squashfs-sysupgrade.bin

Device will reboot, boot U-Boot and then OpenWrt.

Recovery:

In case you need to reinstall OpenWrt if it crashes after U-Boot, there is
a recovery mechanism in OpenWrt to boot the OpenWrt initramfs.
You need to hold the reset button while U-Boot is booting and then it will
boot the OpenWrt initramfs from:
1. USB
2. Networking

In recovery mode U-Boot will light all of the LED-s except for the switch
ones.

In case you want to return to RouterOS, you can simply do that via
Netinstall like on any other MikroTik board.

Credits also go to Serhii Serhieiev <adron@mstnt.com> who origininally
figured out the RouterBoot modification for UART, the missing 10G MVPP2
support in U-Boot as well as the custom aux loader to boot directly via
RouterBoot.

Signed-off-by: Robert Marko <robimarko@gmail.com>

robimarko committed:
mvebu: add support for RB5009UG+S+IN

This patch adds support for Mikrotik RB5009UG+S+IN.

Specifications:
  - SoC: Marvell Armada 7040 (88F7040) - 4 cores, ARMv8 Cortex-A72, 1.4GHz, 64bit
  - RAM: 1024MB DDR4
  - Flash: 16MB SPI NOR flash, 1024MB NAND
  - Ethernet:
  	* Marvell 88E6393X - Amethyst:
  	* one 2.5G RJ45 port via Qualcomm QCA8081 PHY
  	* seven 1G RJ45 ports via built-in PHY-s
  	* one 10G SFP+ cage
  	* All ports share the same 10G switch uplink to the CPU
  - LED: User, SFP, Hdr1, Hdr2
  - Buttons: Reset
  - UART: 115200 8n1 on the MikroTik 16 pin header
  - USB: One USB3 port
  - Power: 24-57 V via
  	* DC jack
  	* 802.3af/at PoE on Ethernet 1
  	* 2-pin terminal on the side

16 Pin header pinout:
1   GND Vcc  RX  ?  GND
   #--------------------#
   |.-. .-. .-. .-. .-. |
   |'-' '-' '-' '-' '-' |
   |.-. .-. .-. .-. .-. |
   |'-' '-' '-' '-' '-' |
   #--------------------#
2   CLK  DO /CS  TX  DI

Do note that the default RouterBoot has disabled UART even when the
required hard-config bit is set to indicate UART support.
Patched RouterBoot must be used if UART is desired.

Also, since ARM64 Linux support does not support in any way appending the
DTB to the kernel image we use mainline U-Boot with added RB5009 support
in order to boot OpenWrt.
MikroTik uses YAFFS to store the boot kernel and we use YAFUT to put U-Boot
as the kernel which RouterBoot then simply boots as an ELF.

Install instructions:

NOTE: In case you are using an existing out of tree version of OpenWrt make
sure to reinstall RouterOS via Netinstall to return the expected partition
layout.

1. Prepare FAT or EXT4 formatted USB drive with OpenWrt initramfs:
* Copy bin/targets/mvebu/cortexa72/openwrt-mvebu-cortexa72-mikrotik_rb5009-initramfs-uImage.itb
to the root of FAT or EXT4 formatted USB drive.
* Plug in the drive to the RB5009 USB port

2. Boot the modified OpenWrt built U-Boot ELF:
u-boot.elf from bin/targets/mvebu/cortexa72/u-boot-rb5009/u-boot.elf

Consult OpenWrt wiki for common instructions on switching to boot from
Ethernet once as well as serving the file:
https://openwrt.org/toh/mikrotik/common

Once U-Boot is booted it will attempt to boot in the following order:
1. NAND
2. USB
3. Network

NAND is expected to fail but USB or Networking need to serve the OpenWrt
initramfs image and after booting it will be accessible from LAN ports
on the default 192.168.1.1 IP with default credentials.

3. Flash modified RouterBoot that enables UART (Optional but recommended):
https://public.robimarko.eu/RB5009/70x0-7.15-uart.fwf

* Copy the file over to the booted OpenWrt initramfs to /tmp
* Run: mtd erase RouterBOOT-primary
* Run: mtd write /tmp/70x0-7.15-uart.fwf RouterBOOT-primary

4. Install U-Boot to boot OpenWrt:
* Copy the u-boot.elf from bin/targets/mvebu/cortexa72/u-boot-rb5009/u-boot.elf
to OpenWrt initramfs to /tmp.
* Run: . /lib/functions.sh
* Run: yafut -d /dev/mtd$(find_mtd_index "YAFFS") -w -i /tmp/u-boot.elf -o kernel -T
This will use yafut to copy the U-Boot as kernel in YAFFS so that RouterBoot boots it.

5. Wipe the NAND UBI partition:
* Run: ubiformat /dev/mtd$(find_mtd_index "ubi") -y
This will prepare the existing RouterOS rootfs partition for OpenWrt.

6. Flash OpenWrt:
* Copy the bin/targets/mvebu/cortexa72/openwrt-mvebu-cortexa72-mikrotik_rb5009-squashfs-sysupgrade.bin
to OpenWrt initramfs to /tmp.
* Run: sysupgrade /tmp/openwrt-mvebu-cortexa72-mikrotik_rb5009-squashfs-sysupgrade.bin

Device will reboot, boot U-Boot and then OpenWrt.

Recovery:

In case you need to reinstall OpenWrt if it crashes after U-Boot, there is
a recovery mechanism in OpenWrt to boot the OpenWrt initramfs.
You need to hold the reset button while U-Boot is booting and then it will
boot the OpenWrt initramfs from:
1. USB
2. Networking

In recovery mode U-Boot will light all of the LED-s except for the switch
ones.

In case you want to return to RouterOS, you can simply do that via
Netinstall like on any other MikroTik board.

Credits also go to Serhii Serhieiev <adron@mstnt.com> who origininally
figured out the RouterBoot modification for UART, the missing 10G MVPP2
support in U-Boot as well as the custom aux loader to boot directly via
RouterBoot.

Link: https://github.com/openwrt/openwrt/pull/15765
Signed-off-by: Robert Marko <robimarko@gmail.com>

robimarko committed:
mvebu: cortex-a72: enable MikroTik platform drivers and NOR variable erase

MikroTik RB5009 will be using advantage of the MikroTik platform drivers,
RouterBoot partition parser and SPI NOR variable erase support.

Link: https://github.com/openwrt/openwrt/pull/15765
Signed-off-by: Robert Marko <robimarko@gmail.com>

robimarko committed:
generic: 6.6: backport Marvell Amethyst SMI GPIO setup

Marvell Amethyst switches use a different SMI GPIO pin setup than other
switches, and since RB5009 uses Amethyst switch and its SMI bus to talk
to QCA8081 lets backport the required fix from kernel 6.9.

Link: https://github.com/openwrt/openwrt/pull/15765
Signed-off-by: Robert Marko <robimarko@gmail.com>

robimarko committed:
mvebu: cortex-a72: enable QCA8081 PHY support

MikroTik RB5009 uses Qualcomm QCA8081 PHY for the 2.5G RJ45 port,
so we need to enable the driver for it.

Link: https://github.com/openwrt/openwrt/pull/15765
Signed-off-by: Robert Marko <robimarko@gmail.com>

robimarko committed:
generic: backport QCA808x possible interfaces fix

QCA808x does not currently fill in the possible_interfaces.

This leads to Phylink not being aware that it supports 2500Base-X as well
so in cases where it is connected to a DSA switch like MV88E6393 it will
limit that port to phy-mode set in the DTS.

That means that if SGMII is used you are limited to 1G only while if
2500Base-X was set you are limited to 2.5G only.

Populating the possible_interfaces fixes this, so lets backport the patches
from kernel 6.9.

This also includes a backport of the Phylink PHY validation series from
kernel 6.8 that allows the use of possible_interfaces.

Link: https://github.com/openwrt/openwrt/pull/15765
Signed-off-by: Robert Marko <robimarko@gmail.com>

robimarko committed:
mvebu: cortex-a72: enable ARM SBSA Generic Watchdog

Marvell 70x0 and 80x0 both have ARM SBSA Generic Watchdog built-in,
so lets enable the required driver for them.

Link: https://github.com/openwrt/openwrt/pull/15765
Signed-off-by: Robert Marko <robimarko@gmail.com>

robimarko committed:
mvebu: cortex-a72: enable U-Boot NVMEM driver

In order to not have to ship envtools configuration per board, we can
instead rely on the kernel U-Boot environment NVMEM driver through which
envtools can read/write the environement.

Since size difference is negligeble and this subtarget has rather large
storage regardless, enable it by default.

Link: https://github.com/openwrt/openwrt/pull/15765
Signed-off-by: Robert Marko <robimarko@gmail.com>

robimarko committed:
uboot-mvebu: add support for MikroTik RB5009

MikroTik RB5009 uses RouterBoot as its bootloader like all MikroTik devices
running RouterOS, meaning that its not FIT compatible and can only boot
ELF images.

Now this is not so much of an issue on ARM or MIPS since kernel supports
appending DTB-s to it (Or we patch the kernel to embed it), but on ARM64
there is intentionally no such support.

RouterBoot will pass a DTB, but its the broken MikroTik one which is a
modified reference DTB and incorrect in more places than its valid so we
cannot use it to boot our kernel.

Thus, the solution is to use an intermediary loader and luckily for us
Armada 7040 is well supported in U-Boot which makes it a great option since
it supports anything that we will ever need to boot.

Upstream U-Boot currently requires the Armada boards to be converted to
OF_UPSTREAM before adding anything new and this requires updating all of
the drivers to accomodate the Linux DTS, while I plan to do this eventually
we will need to keep this board downstream for now.

Most stuff is supported in U-Boot, including networking since the switch
is preconfigured by RouterBoot.

A custom environment is used to try and boot from the following devices:
1. NAND (UBI)
2. USB
3. Networking

If NAND boot fails then U-Boot will attempt to boot OpenWrt initramfs from
USB or via networking.

There is a manual recovery mechanism implemented where if the reset button
is held when U-Boot is booting it will try to boot OpenWrt initramfs from:
1. USB
2. Networking

When U-Boot is in recovery mode it will light all of the LED-s except the
switch ones.

Link: https://github.com/openwrt/openwrt/pull/15765
Signed-off-by: Robert Marko <robimarko@gmail.com>

robimarko committed:
mvebu: cortex-a72: enable MikroTik NVMEM layout driver

RB5009 will take advantage of the driver to get MAC adress.

Link: https://github.com/openwrt/openwrt/pull/15765
Signed-off-by: Robert Marko <robimarko@gmail.com>

john-tho commented on target/linux/mvebu/files-6.6/arch/arm64/boot/dts/marvell/armada-7040-rb5009.dts:
```diff
@@ -0,0 +1,361 @@
+			};
+
+			partition@b0000 {
+				label = "RouterBOOT";
```
Could we please make it more clear that this partition is primary bootloader (only)? Perhaps: `bootloader2`?
There are many other `RouterBoot|routerboot` partitions in tree, and they are all both bootloaders+configs.

<details><summary><code>git grep -B2 -i '"routerboot";' master -- target/linux/{ath79,ipq40xx,ramips}</code></summary>

```
master:target/linux/ath79/dts/ar9342_mikrotik_routerboard-911g.dtsi-
master:target/linux/ath79/dts/ar9342_mikrotik_routerboard-911g.dtsi-			partition@0 {
master:target/linux/ath79/dts/ar9342_mikrotik_routerboard-911g.dtsi:				label = "routerboot";
--
master:target/linux/ath79/dts/ar9344_mikrotik_routerboard-16m-nor.dtsi-
master:target/linux/ath79/dts/ar9344_mikrotik_routerboard-16m-nor.dtsi-			partition@0 {
master:target/linux/ath79/dts/ar9344_mikrotik_routerboard-16m-nor.dtsi:				label = "RouterBoot";
--
master:target/linux/ath79/dts/ar9344_mikrotik_routerboard-951x-2hnd.dtsi-
master:target/linux/ath79/dts/ar9344_mikrotik_routerboard-951x-2hnd.dtsi-			partition@0 {
master:target/linux/ath79/dts/ar9344_mikrotik_routerboard-951x-2hnd.dtsi:				label = "routerboot";
--
master:target/linux/ath79/dts/ar9344_mikrotik_routerboard-sxt-5nd-r2.dts-
master:target/linux/ath79/dts/ar9344_mikrotik_routerboard-sxt-5nd-r2.dts-			partition@0 {
master:target/linux/ath79/dts/ar9344_mikrotik_routerboard-sxt-5nd-r2.dts:				label = "RouterBoot";
--
master:target/linux/ath79/dts/qca9533_mikrotik_routerboard-16m.dtsi-
master:target/linux/ath79/dts/qca9533_mikrotik_routerboard-16m.dtsi-			partition@0 {
master:target/linux/ath79/dts/qca9533_mikrotik_routerboard-16m.dtsi:				label = "RouterBoot";
--
master:target/linux/ath79/dts/qca9556_mikrotik_routerboard-wap-g-5hact2hnd.dts-
master:target/linux/ath79/dts/qca9556_mikrotik_routerboard-wap-g-5hact2hnd.dts-			partition@0 {
master:target/linux/ath79/dts/qca9556_mikrotik_routerboard-wap-g-5hact2hnd.dts:				label = "RouterBoot";
--
master:target/linux/ath79/dts/qca9558_mikrotik_routerboard-92x.dtsi-
master:target/linux/ath79/dts/qca9558_mikrotik_routerboard-92x.dtsi-			partition@0 {
master:target/linux/ath79/dts/qca9558_mikrotik_routerboard-92x.dtsi:				label = "routerboot";
--
master:target/linux/ath79/dts/qca9558_mikrotik_routerboard-96x.dtsi-
master:target/linux/ath79/dts/qca9558_mikrotik_routerboard-96x.dtsi-			partition@0 {
master:target/linux/ath79/dts/qca9558_mikrotik_routerboard-96x.dtsi:				label = "RouterBoot";
--
master:target/linux/ipq40xx/files-6.6/arch/arm/boot/dts/qcom/qcom-ipq4018-cap-ac.dts-				#address-cells = <1>;
master:target/linux/ipq40xx/files-6.6/arch/arm/boot/dts/qcom/qcom-ipq4018-cap-ac.dts-				#size-cells = <1>;
master:target/linux/ipq40xx/files-6.6/arch/arm/boot/dts/qcom/qcom-ipq4018-cap-ac.dts:				label = "RouterBoot";
--
master:target/linux/ipq40xx/files-6.6/arch/arm/boot/dts/qcom/qcom-ipq4018-hap-ac2.dts-				#address-cells = <1>;
master:target/linux/ipq40xx/files-6.6/arch/arm/boot/dts/qcom/qcom-ipq4018-hap-ac2.dts-				#size-cells = <1>;
master:target/linux/ipq40xx/files-6.6/arch/arm/boot/dts/qcom/qcom-ipq4018-hap-ac2.dts:				label = "RouterBoot";
--
master:target/linux/ipq40xx/files-6.6/arch/arm/boot/dts/qcom/qcom-ipq4018-sxtsq-5-ac.dts-				#address-cells = <1>;
master:target/linux/ipq40xx/files-6.6/arch/arm/boot/dts/qcom/qcom-ipq4018-sxtsq-5-ac.dts-				#size-cells = <1>;
master:target/linux/ipq40xx/files-6.6/arch/arm/boot/dts/qcom/qcom-ipq4018-sxtsq-5-ac.dts:				label = "RouterBoot";
--
master:target/linux/ipq40xx/files-6.6/arch/arm/boot/dts/qcom/qcom-ipq4018-wap-ac.dtsi-				#address-cells = <1>;
master:target/linux/ipq40xx/files-6.6/arch/arm/boot/dts/qcom/qcom-ipq4018-wap-ac.dtsi-				#size-cells = <1>;
master:target/linux/ipq40xx/files-6.6/arch/arm/boot/dts/qcom/qcom-ipq4018-wap-ac.dtsi:				label = "RouterBoot";
--
master:target/linux/ipq40xx/files-6.6/arch/arm/boot/dts/qcom/qcom-ipq4019-hap-ac3-lte6-kit.dts-			partition@80000 {
master:target/linux/ipq40xx/files-6.6/arch/arm/boot/dts/qcom/qcom-ipq4019-hap-ac3-lte6-kit.dts-				compatible = "mikrotik,routerboot-partitions";
master:target/linux/ipq40xx/files-6.6/arch/arm/boot/dts/qcom/qcom-ipq4019-hap-ac3-lte6-kit.dts:				label = "RouterBoot";
--
master:target/linux/ipq40xx/files-6.6/arch/arm/boot/dts/qcom/qcom-ipq4019-hap-ac3.dts-				#address-cells = <1>;
master:target/linux/ipq40xx/files-6.6/arch/arm/boot/dts/qcom/qcom-ipq4019-hap-ac3.dts-				#size-cells = <1>;
master:target/linux/ipq40xx/files-6.6/arch/arm/boot/dts/qcom/qcom-ipq4019-hap-ac3.dts:				label = "RouterBoot";
--
master:target/linux/ipq40xx/files-6.6/arch/arm/boot/dts/qcom/qcom-ipq4019-lhgg-60ad.dts-				#address-cells = <1>;
master:target/linux/ipq40xx/files-6.6/arch/arm/boot/dts/qcom/qcom-ipq4019-lhgg-60ad.dts-				#size-cells = <1>;
master:target/linux/ipq40xx/files-6.6/arch/arm/boot/dts/qcom/qcom-ipq4019-lhgg-60ad.dts:				label = "RouterBoot";
--
master:target/linux/ramips/dts/mt7621_mikrotik.dtsi-
master:target/linux/ramips/dts/mt7621_mikrotik.dtsi-			partition@0 {
master:target/linux/ramips/dts/mt7621_mikrotik.dtsi:				label = "RouterBoot";
```

</details>

robimarko replied:
All of those other occurrences are basically wrong and should be renamed as RouterBOOT is the bootloader.

john-tho replied:
> RouterBOOT is the bootloader

Technicalities, but, I'm not so sure. primary bootloader is the fwf file from RouterOS squashfs, less header. Backup bootloader has extras.
There is a stub somewhere that checks if NOR soft config is set for backup bootloader, or if reset button is held while power applied. My guess is that is part of the extras in backup bootloader.
If primary bootloader was _the bootloader_, device would still work if you erased backup bootloader.

robimarko replied:
Oh, I think we misunderstood each other.
This is just the primary instance of RouterBOOT, there is a backup one as well after ATF.
On this device I think they put FWF decompression and everything else directly into ATF.

> What would you suggest we rename our historic in-tree routerboot partitions to? Being all the space used by RouterBoot (bootloaders, plus the settings they need to function)?

Well, I just named the "container" partition "MikroTik" as I dont have a good name to use for it.

> I would be happy if you gave RouterBoot a suffix here?

`RouterBOOT-primary` and `RouterBOOT-backup`?

john-tho commented:
Huge thanks for getting this working well.


>     1. Boot the modified OpenWrt initramfs(Patched to work with RouterBoot):
>        https://public.robimarko.eu/RB5009/openwrt-mvebu-cortexa72-mikrotik_rb5009-initramfs-fit-uImage.elf
> 

Can we clarify how this is built, and that it is only needed to load kernel (with our) DTB? As alternatives, can use u-boot.elf built with `CONFIG_BOOTCOMMAND` set to `dhcp; tftpboot; bootm` as a separate rb5009 u-boot netboot build, or use u-boot.elf with serial or netconsole. I also saw _U-Boot Standard Boot_, and considered that with priority usb (to be able to force work around a broken OpenWrt on nand), nand, network.


>     3. Install U-Boot to boot OpenWrt:
> 
> 
>     * Copy the u-boot.elf from bin/targets/mvebu/cortexa72/u-boot-rb5009/u-boot.elf
>       to OpenWrt initramfs to /tmp.
> 
>     * Run: yafut -d /dev/mtd0 -w -i /tmp/u-boot.elf -o kernel -T
>       This will use yafut to copy the U-Boot as kernel in YAFFS so that RouterBoot boots it.
> 
> 
>     4. Wipe the NAND UBI partition:
> 
> 
>     * Run: cat /proc/mtd | grep ubi
>       Note the mtdN number.
> 
>     * Run: ubiformat /dev/mtd1 -y
>       Make sure that mtd1 is the correct partition.

Could we use partition names in these commands? We know that people run these commands without checking, and inevitably someone will try this from an older image with different partition layout. We have a function:
https://github.com/openwrt/openwrt/blob/73a9f9f857947bc63fef1aefb7e2e4b906fb9d2b/package/base-files/files/lib/functions.sh#L418

possibly:
```
. /lib/functions.sh

yafut -d /dev/mtd$(find_mtd_index "YAFFS") -w -i /tmp/u-boot.elf -o kernel -T
#…
ubiformat /dev/mtd$(find_mtd_index "ubi") -y
```

Cheers

robimarko commented:
I will answer point by point.

> 1. Boot the modified OpenWrt initramfs(Patched to work with RouterBoot):
>    https://public.robimarko.eu/RB5009/openwrt-mvebu-cortexa72-mikrotik_rb5009-initramfs-fit-uImage.elf
> 
> Can we clarify how this is built, and that it is only needed to load kernel (with our) DTB? As alternatives, can use u-boot.elf built with CONFIG_BOOTCOMMAND set to dhcp; tftpboot; bootm as a separate rb5009 u-boot netboot build, or use u-boot.elf with serial or netconsole. I also saw U-Boot Standard Boot, and considered that with priority usb (to be able to force work around a broken OpenWrt on nand), nand, network.

This is built with the custom Aux loader by Serhii, you can find the branch from which it was built here:
https://github.com/robimarko/openwrt/tree/RB5009-upstreaming-AUX-loader

It is used as a practical way to be able to boot a full OpenWrt image via RouterBOOT only for installation purposes,  and while it is practical I am not a fan of it as its a custom image.

Yes, one can use the OpenWrt built `u-boot.elf` and boot that via RouterBOOT instead and then use it to boot the standard OpenWrt initramfs image.
I have only briefly looked at `bootstd` as it does not support NOR and NAND device as far as I could tell, and even if it did it would require us to ship the bootflow recipe(Essentially the bootscript) somewhere.

Netconsole is an option, I think that once trying to boot from NAND/UBI fails it will just fallback to console and thus one could use that or UART instead of the custom OpenWrt image.

I agree that this is a good place to invest some time as having mainline U-Boot gives us quite some flexibility.

robimarko commented:
@john-tho I have updated the U-Boot to allow completely skipping the patched initramfs as well as instructions, also included the updated instructions to find the MTD index.

robimarko committed:
treewide: refresh patches

These need to be refreshed, probably the generic backports affected them.

Link: https://github.com/openwrt/openwrt/pull/15765
Signed-off-by: Robert Marko <robimarko@gmail.com>

john-tho commented:
Thank you for the explanations and changes. I think if we link to binaries in commit message, we should briefly describe what they do/are, so that someone else has a chance if links are inaccessible to them.

I have not tested these u-boot changes, but they make sense. If we get into a situation where the fit image is okay on nand, but it fails to start OpenWrt, doing `ubi, usb, net` we would be stuck. I don't think worth churning NOR/NAND to setup a bootcount incremented in u-boot and reset in OpenWrt. Could prioritise USB (but extra boot time for the setup and scan, and might be fragile), or (not sure how much extra compute it would take to build, but) I am leaning towards building a second u-boot for rb5009-netboot that prioritises USB then tftpboot. Or just mention somewhere that this workaround (or adron's loader) could be needed.

I think we will not be able to set u-boot env settings from fresh on OpenWrt as is. The `u_boot_env` driver bails out if an invalid CRC, then `fw_env.c` fails to find the nvmem bus for `u-boot,env`. Could either set env if unset in u-boot, or carry the `fw_env.config` in OpenWrt.

Cheers

john-tho commented on target/linux/mvebu/files-6.6/arch/arm64/boot/dts/marvell/armada-7040-rb5009.dts:
```diff
@@ -0,0 +1,361 @@
+};
+
+&cp0_gpio2 {
+	enable-usb-power {
+		gpio-hog;
+		gpios = <23 GPIO_ACTIVE_HIGH>;
+		output-high;
+		line-name = "enable USB power";
+	};
+
+	enable-leds-power {
+		gpio-hog;
+		gpios = <27 GPIO_ACTIVE_HIGH>;
+		output-high;
+		line-name = "enable LED-s power";
+	};
```
Hogging power switches has always annoyed me.
I might see if I can get [`regulator-output`](https://elixir.bootlin.com/linux/latest/B/ident/regulator-output) to work for these?
I looked at (USB power switch regulator) [some time ago for a different device](https://github.com/openwrt/openwrt/issues/10904#issuecomment-1312645744)

Would also like to see the device-specific gpios named in `gpio-line-names`. Will look at that too.

john-tho replied:
userspace regulator works for me, allows control, and uses upstream driver.
guessed LED voltage here. Not concerned if you don't use it, or don't need attribution if you do.

```patch
commit 60205e998580dd23d8bf58da058d2e952a52fd26
Author: John Thomson <git@johnthomson.fastmail.com.au>
Date:   Sun Jun 23 08:42:59 2024 +1000

    mvebu: rb5009: userspace regulator for power switches
    
    Use userspace controlled regulators to manage USB port power and LED
    power switches.
    
    echo "disabled">/sys/bus/platform/drivers/reg-userspace-consumer/output-usb-port-power/state
    sleep 1
    echo "enabled">/sys/bus/platform/drivers/reg-userspace-consumer/output-usb-port-power/state
    echo "disabled">/sys/bus/platform/drivers/reg-userspace-consumer/output-led-power/state
    
    Signed-off-by: John Thomson <git@johnthomson.fastmail.com.au>

diff --git a/target/linux/mvebu/cortexa72/config-6.6 b/target/linux/mvebu/cortexa72/config-6.6
index de2cec6f26ba..19ca2b29d1b1 100644
--- a/target/linux/mvebu/cortexa72/config-6.6
+++ b/target/linux/mvebu/cortexa72/config-6.6
@@ -102,6 +102,7 @@ CONFIG_QUEUED_SPINLOCKS=y
 CONFIG_RAS=y
 # CONFIG_RAVE_SP_CORE is not set
 CONFIG_REGULATOR_GPIO=y
+CONFIG_REGULATOR_USERSPACE_CONSUMER=y
 # CONFIG_RODATA_FULL_DEFAULT_ENABLED is not set
 CONFIG_SENSORS_IEI_WT61P803_PUZZLE_HWMON=y
 CONFIG_SERIAL_DEV_BUS=y
diff --git a/target/linux/mvebu/files-6.6/arch/arm64/boot/dts/marvell/armada-7040-rb5009.dts b/target/linux/mvebu/files-6.6/arch/arm64/boot/dts/marvell/armada-7040-rb5009.dts
index 886e87222b3a..9cc5ce22b98c 100644
--- a/target/linux/mvebu/files-6.6/arch/arm64/boot/dts/marvell/armada-7040-rb5009.dts
+++ b/target/linux/mvebu/files-6.6/arch/arm64/boot/dts/marvell/armada-7040-rb5009.dts
@@ -76,6 +76,36 @@
 		tx-disable-gpios = <&cp0_gpio1 5 GPIO_ACTIVE_HIGH>;
 		rate-select0-gpios = <&cp0_gpio1 3 GPIO_ACTIVE_HIGH>;
 	};
+
+	reg_usb_port: regulator-usb {
+		compatible = "regulator-fixed";
+		regulator-name = "USB-power";
+		regulator-min-microvolt = <5000000>;
+		regulator-max-microvolt = <5000000>;
+		gpio = <&cp0_gpio2 23 GPIO_ACTIVE_HIGH>;
+		enable-active-high;
+		regulator-boot-on;
+	};
+
+	output-usb-port-power {
+		compatible = "regulator-output";
+		vout-supply = <&reg_usb_port>;
+	};
+
+	reg_leds: regulator-leds {
+		compatible = "regulator-fixed";
+		regulator-name = "LED-power";
+		regulator-min-microvolt = <3300000>;
+		regulator-max-microvolt = <3300000>;
+		gpio = <&cp0_gpio2 27 GPIO_ACTIVE_HIGH>;
+		enable-active-high;
+		regulator-boot-on;
+	};
+
+	output-led-power {
+		compatible = "regulator-output";
+		vout-supply = <&reg_leds>;
+	};
 };
 
 &uart0 {
@@ -177,22 +207,6 @@
 	};
 };
 
-&cp0_gpio2 {
-	enable-usb-power {
-		gpio-hog;
-		gpios = <23 GPIO_ACTIVE_HIGH>;
-		output-high;
-		line-name = "enable USB power";
-	};
-
-	enable-leds-power {
-		gpio-hog;
-		gpios = <27 GPIO_ACTIVE_HIGH>;
-		output-high;
-		line-name = "enable LED-s power";
-	};
-};
-
 &cp0_usb3_1 {
 	status = "okay";
 };
```

robimarko replied:
Thanks, I wired up the USB regulator as the supply to the USB connector so the host controller driver will enable it automatically.

For LED-s I used your userspace regulator method.

While I was here, noticed that SERDES lanes were not described in Linux nor UTMI PHY was enabled so sorted that out.

robimarko commented:
> Thank you for the explanations and changes. I think if we link to binaries in commit message, we should briefly describe what they do/are, so that someone else has a chance if links are inaccessible to them.

The only binary being linked is the patched RouterBOOT, and its explained what the purpose is.
As for the link getting lost, that's always a possibility though I will keep my public host live for as long as possible.
It's not feasible to try and write instructions on how to patch the RouterBoot for me.

> I have not tested these u-boot changes, but they make sense. If we get into a situation where the fit image is okay on nand, but it fails to start OpenWrt, doing `ubi, usb, net` we would be stuck. I don't think worth churning NOR/NAND to setup a bootcount incremented in u-boot and reset in OpenWrt. Could prioritise USB (but extra boot time for the setup and scan, and might be fragile), or (not sure how much extra compute it would take to build, but) I am leaning towards building a second u-boot for rb5009-netboot that prioritises USB then tftpboot. Or just mention somewhere that this workaround (or adron's loader) could be needed. Edit: I guess we could use the reset button?

I tested them and they work in the documented order, however you have a valid point regarding the case where it boots OpenWrt but then crashes.

I have wired up the reset button for the recovery, so if its held while U-Boot is booting it will try booting from USB and then from networking, this allows recovering.

I will also look into wiring a LED for indicating the status.
 
> I think we will not be able to set u-boot env settings from fresh on OpenWrt as is. The `u_boot_env` driver bails out if an invalid CRC, then `fw_env.c` fails to find the nvmem bus for `u-boot,env`. Could either set env if unset in u-boot, or carry the `fw_env.config` in OpenWrt.
> 
> Cheers

Yes, this is intentional only for those that actually stop U-Boot and save the env, as we dont want to ship the envtools config so that users can override the default env that easy.

robimarko commented:
@dchard I have pushed the U-Boot changes.

I would really love to get this merged, we can always tweak U-Boot or other stuff later.

john-tho commented:
> @john-tho I have pushed the U-Boot changes.
> 
> I would really love to get this merged, we can always tweak U-Boot or other stuff later.

No blockers here. Works for me, and you have addressed all my feedback, thank you!
Great work. It is amazing working with a _real_ (extensible) bootloader!

robimarko commented:
Thanks! Rebased on top of main and merged!

john-tho commented on target/linux/mvebu/files-6.6/arch/arm64/boot/dts/marvell/armada-7040-rb5009.dts:
```diff
@@ -0,0 +1,384 @@
+	clock-frequency = <100000>;
+};
+
+&cp0_mdio {
+	status = "okay";
+};
+
```
Noticed this is duplicated below (where it includes switch definition)

john-tho commented on target/linux/mvebu/files-6.6/arch/arm64/boot/dts/marvell/armada-7040-rb5009.dts:
```diff
@@ -0,0 +1,384 @@
+			#address-cells = <1>;
+			#size-cells = <0>;
+
+			qca8081: qca8081@0 {
+				reg = <0>;
+			};
```
No idea how _valid_ it is but if I `active-low` the LED, I seem to get matching behaviour with green and amber (does not come on) between RouterBoot, u-boot, and OpenWrt (but appears to spend some seconds inverted while booting before this gets applied).
```
			qca8081: qca8081@0 {
				compatible = "ethernet-phy-id004d.d101";
				reg = <0>;

				leds {
					#address-cells = <1>;
					#size-cells = <0>;

					led@0 {
						reg = <0>;
						color = <LED_COLOR_ID_GREEN>;
						function = LED_FUNCTION_LAN;
						default-state = "keep";
						active-low;
					};


					led@1 {
						reg = <1>;
						color = <LED_COLOR_ID_AMBER>;
						function = LED_FUNCTION_LAN;
						default-state = "keep";
						active-low;
					};
				};
			};
```

robimarko replied:
This should be valid, that is why Ansuel added the active-low cause some actually use it like that.

I need to get around to RB5009 (Hopefully tomorrow) to get this sorted out.

damex replied:
is it related to rb5009 sfp port led never lighting up? i tried sfp+ dac, 1g, 2.5g and 10g copper modules - no light come up but module is detected by openwrt

john-tho replied:
> is it related to rb5009 sfp port led never lighting up? i tried sfp+ dac, 1g, 2.5g and 10g copper modules - no light come up but module is detected by openwrt

No, not related to SFP LED. You need to define a trigger for it, see `cat /sys/class/leds/green:sfp/trigger`.
Example:
```base
root@OpenWrt-rb5009:~# grep -A10 sfp /etc/config/system 
config led 'sfp'
        option name 'SFP'
        option sysfs 'green:sfp'
        option trigger 'netdev'
        option dev 'sfp'
        option mode 'tx rx'
        option interval '50'
```
```base
/etc/init.d/led restart
```

damex replied:
thanks, that solves sfp port.

been playing around with rb5009 and noticed 2.5gbe port light always being on without cable being plugged in. light gets proper triggering when there is activity. is that the issue with light above? :)

i have two rb5009 and both are the same.
![photo_2024-07-11_18-50-31](https://github.com/openwrt/openwrt/assets/450390/fb3e55af-0dc5-49d9-97ec-272bf67eea49)

sppmasterspp replied:
> been playing around with rb5009 and noticed 2.5gbe port light always being on without cable being plugged in.

That issue has been addressed with latest commits.
